### PR TITLE
executor: support xapi expression push down.

### PR DESCRIPTION
**Pull request by coocood**

### CODE REVIEW

File: executor/executor_xapi.go [hunk 1]
```diff
@@ -460,3 +487,120 @@ type int64Slice []int64
 func (p int64Slice) Len() int           { return len(p) }
 func (p int64Slice) Less(i, j int) bool { return p[i] < p[j] }
 func (p int64Slice) Swap(i, j int)      { p[i], p[j] = p[j], p[i] }
+
+// exprToPBExpr converts an ast.ExprNode to a tipb.Expr, if not supported, nil will be returned.
+func exprToPBExpr(client kv.Client, expr ast.ExprNode, fields []*ast.ResultField) *tipb.Expr {
```
Comment (shenli) [inline executor/executor_xapi.go]: Why not use a visitor?

File: executor/builder.go [hunk 2]
```diff
@@ -198,12 +196,11 @@ func (b *executorBuilder) buildIndexScan(v *plan.IndexScan) Executor {
...
 		}
-		where := conditionsToPBExpression(v.FilterConditions...)
-		if xapi.SupportExpression(client, where) {
+		where, remained := conditionsToPBExpr(client, v.FilterConditions, v.Fields())
+		if where != nil {
 			e.where = where
-			return e
 		}
-		return b.buildFilter(e, v.FilterConditions)
+		return b.buildFilter(e, remained)
```
Comment (shenli) [inline executor/builder.go]: If remained == nil, we do not need the filter?

Comment (shenli) [inline executor/builder.go]: Same as above

File: executor/executor_xapi.go [hunk 3]
```diff
@@ -264,12 +278,25 @@ func (e *XSelectIndexExec) doTableRequest(txn kv.Transaction, handles []int64) (
...
 		selTableReq.Ranges = append(selTableReq.Ranges, pbRange)
 	}
-	selTableReq.Where = conditionsToPBExpression(e.indexPlan.FilterConditions...)
+	selTableReq.Where = e.where
 	return xapi.Select(txn.GetClient(), selTableReq, 10)
 }
 
-func conditionsToPBExpression(expr ...ast.ExprNode) *tipb.Expr {
-	return nil
+func conditionsToPBExpr(client kv.Client, exprs []ast.ExprNode, fields []*ast.ResultField) (pbexpr *tipb.Expr,
```
Comment (shenli) [inline executor/executor_xapi.go]: Please add a few comment for this function.

File: xapi/xeval/eval.go [hunk 4]
```diff
@@ -271,13 +271,21 @@ func (e *Evaluator) evalTwoBoolChildren(expr *tipb.Expr) (leftBool, rightBool in
 	if err != nil {
 		return 0, 0, errors.Trace(err)
 	}
-	leftBool, err = left.ToBool()
-	if err != nil {
-		return 0, 0, errors.Trace(err)
+	if left.Kind() == types.KindNull {
+		leftBool = compareResultNull
```
Comment (BusyJay) [inline xapi/xeval/eval.go]: Why not just use 0? I don't think `and` or `or` is comparison.

File: executor/builder.go (hunk 2 above)
Comment (coocood) [inline executor/builder.go]: `buildFilter` function will check if remained is empty and returns e instead.

File: executor/executor_xapi.go (hunk 1 above)
Comment (coocood) [inline executor/executor_xapi.go]: Use a function to convert expression is simpler to implement.

File: xapi/xeval/eval.go (hunk 4 above)
Comment (coocood) [inline xapi/xeval/eval.go]: in case `evalTwoBoolChildren` being used in expressions other and 'AND', 'OR'.

File: executor/executor_xapi.go [hunk 5]
```diff
@@ -204,22 +206,34 @@ func (e *XSelectIndexExec) doRequest() error {
...
 	return nil
 }
 
+type rowsSorter struct {
+	order map[int64]int
+	rows  []*Row
+}
+
+func (s *rowsSorter) Less(i, j int) bool {
+	x := s.order[s.rows[i].RowKeys[0].Handle]
```
Comment (shenli) [inline executor/executor_xapi.go]: How about handle not in s.order?

**Comment by shenli** [comment]
LGTM

File: executor/executor_xapi.go [hunk 6]
```diff
@@ -460,3 +487,120 @@ type int64Slice []int64
...
+	}
+	switch column.Refer.Expr.GetType().Tp {
+	case mysql.TypeBit, mysql.TypeSet, mysql.TypeEnum, mysql.TypeDecimal, mysql.TypeNewDecimal, mysql.TypeGeometry,
+		mysql.TypeDate, mysql.TypeNewDate, mysql.TypeDatetime, mysql.TypeDuration, mysql.TypeTimestamp, mysql.TypeYear:
+		return nil
+	}
+	matched := false
+	for _, f := range fields {
+		if f.Column.ID == column.Refer.Column.ID {
+			matched = true
```
Comment (zimulala) [inline executor/executor_xapi.go]: Here add `break` ?

File: executor/executor_xapi.go (hunk 5 above)
Comment (coocood) [inline executor/executor_xapi.go]: That's not possible.

**Comment by coocood** [comment]
@zimulala PTAL

File: store/localstore/local_region.go [hunk 7]
```diff
@@ -144,12 +189,39 @@ func (rs *localRegion) getRowByHandle(txn kv.Transaction, tid, handle int64, col
...
 			key := tablecodec.EncodeColumnKey(tid, handle, col.GetColumnId())
-			data, err := txn.Get(key)
+			data, err1 := txn.Get(key)
+			if err1 != nil {
+				return nil, errors.Trace(err1)
+			}
+			row.Data = append(row.Data, data...)
+		}
+	}
+	if where != nil {
```
Comment (zimulala) [inline store/localstore/local_region.go]: "if where == nil {return row, nil}" is better ?

**Comment by zimulala** [comment]
LGTM

**Pull request merged**